### PR TITLE
More comprehensive "make clean" action

### LINKED ISSUE #337: Make "clean" action more robust
The current dev env supports a `make clean` action, which attempts to purge SDW-specific configurations. There are a few exceptions, however, which linger after the clean:

* `sys-firewall` DNF repo customizations (in `/rw/config/` and `/etc/`)
* dom0 RPC policies
* legacy VM names (e.g. `sd-export-usb`, see https://github.com/freedomofpress/securedrop-workstation/pull/323#pullrequestreview-316576701)
* `qubes-prefs` such as `default_dispvm` (#312)

In particular, the `qubes-prefs` issue _prevents_ removal of the `sd-svs-disp{,-template}` VMs, so we should consider reverting that immediately until this issue is resolved.

### LINKED ISSUE #350: make clean does not remove sd-svs-disp and sd-svs-disp template
This is because is it set as default dispvm for non-SecureDrop Qubes:

```
[user@dom0 securedrop-workstation]$ qvm-remove sd-svs-disp
This will completely remove the selected VM(s)...
  sd-svs-disp
Are you sure? [y/N] y
VM sd-svs-disp cannot be removed. It is in use as:
 - global property default_dispvm
```

As part of the clean operation, we should first set the default DispVM option to the Qubes-default (`Fedora-30-dvm`) prior to deleting `sd-svs-disp(-template)`

### PR DESCRIPTION
**Pull request by conorsch**
Expands the "make clean" logic to purge configuration files on dom0 and sys-firewall created during the installation of SDW. Now "make clean" will restore the system state quite nearly to the point prior to installing in the first place. One notable exception is that it won't purge the TemplateVM RPM packages, because re-fetching those every time would be time- and bandwidth-intensive.

Relevant issues:

* Closes #337
* Closes #350 
* Refs #349 (test cleanup)
* Refs #339 (use of `update.qubes-vm` at install time)

The clean-it-like-you-mean-it approach will be particularly useful as we tinker with strategies for improving the update mechanisms (#341, #238).

There are a few changes in here, particularly the top file consolidation, that result in a large diff, but a substantially more readable configuration over time. While not the best metric, the changes reduce both the number of files and the number of LoC to maintain. If you weight the transition from Bash to Python, then the results are even better.

### LoC Before
```
 $ cloc . --force-lang YAML,sls --force-lang YAML,top 
      89 text files.
      89 unique files.                              
      29 files ignored.

github.com/AlDanial/cloc v 1.81  T=0.02 s (3037.3 files/s, 142505.4 lines/s)
--------------------------------------------------------------------------------
Language                      files          blank        comment           code
--------------------------------------------------------------------------------
YAML                             43            145            272            835
Python                           16            179            107            573
Markdown                          2            240              0            562
Bourne Again Shell                9             71             86            209
make                              1             35              9            117
JSON                              1              0              0             10
Dockerfile                        1              2              1              9
Bourne Shell                      1              1              3              6
--------------------------------------------------------------------------------
SUM:                             74            673            478           2321
--------------------------------------------------------------------------------
```
### LoC after
```
 $ cloc . --force-lang YAML,sls --force-lang YAML,top 
      70 text files.
      70 unique files.                              
      29 files ignored.

github.com/AlDanial/cloc v 1.81  T=0.02 s (2380.2 files/s, 147876.9 lines/s)
--------------------------------------------------------------------------------
Language                      files          blank        comment           code
--------------------------------------------------------------------------------
YAML                             25            131            238            830
Python                           17            193            121            627
Markdown                          2            240              0            562
Bourne Again Shell                7             63             75            160
make                              1             34              9            102
JSON                              1              0              0             10
Dockerfile                        1              2              1              9
Bourne Shell                      1              1              3              6
--------------------------------------------------------------------------------
SUM:                             55            664            447           2306
--------------------------------------------------------------------------------
```

### Testing

```
make clean # to run the old clean logic from master
qvm-ls --tags sd-workstation # observe sd-svs-disp still hanging around
make clone
make clean # to run the new clean logic from this PR
qvm-ls --tags sd-workstation # observe only RPM-installed template remains
make all
make test
```

### CODE REVIEW

File: dom0/sd-clean-all.sls [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    - names:
+      - qvm-run sys-firewall 'sudo rm -f /rw/config/RPM-GPG-KEY-securedrop-workstation-test'
+      - qvm-run sys-firewall 'sudo rm -f /rw/config/sd-copy-rpm-repo-pubkey.sh'
+      - qvm-run sys-firewall 'sudo rm -f /etc/pki/rpm-gpg/RPM-GPG-KEY-securedrop-workstation-test'
+      - qvm-run sys-firewall 'sudo perl -pi -E "s#^/rw/config/sd-copy-rpm-repo-pubkey.sh##" /rw/config/rc.local'
+
+sd-cleanup-rpc-mgmt-policy:
+  file.replace:
+    - names:
+      - /etc/qubes-rpc/policy/qubes.VMShell
```
Comment (emkll) [inline dom0/sd-clean-all.sls]: these policies do not exist on my machine. How/when are these created? qubesctl?

File: dom0/sd-clean-all.sls [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      - /etc/qubes-rpc/policy/qubes.VMRootShell
+    - repl: ''
+    - pattern: '^disp-mgmt-sd-\w+\s+sd-\w+\s+allow,user=root'
+
+{% set sdw_customized_rpc_files = salt['cmd.shell']('grep -rl "BEGIN securedrop-workstation" /etc/qubes-rpc/ | cat').splitlines() %}
```
Comment (emkll) [inline dom0/sd-clean-all.sls]: Awesome! There's one edge case that we could. When vim or other editor leaves a swap file (like it did on my system), the file is viewed as a binary and the `make clean` action will fail. This is obviously an edge case (and deleting that file resolved), see output below. Perhaps grepping for strings that don't start with a dot is sufficient here, but IMO not required for merge.


```
----------
          ID: sd-cleanup-rpc-policy-grants
    Function: file.replace
        Name: /etc/qubes-rpc/policy/.qubes.Filecopy.swp
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/site-packages/salt/state.py", line 1837, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1794, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python2.7/site-packages/salt/states/file.py", line 3934, in replace
                  ignore_if_missing=ignore_if_missing)
                File "/usr/lib/python2.7/site-packages/salt/modules/file.py", line 2035, in replace
                  .format(path)
              SaltInvocationError: Cannot perform string replacements on a binary file: /etc/qubes-rpc/policy/.qubes.Filecopy.swp
     Started: 16:16:15.438709
    Duration: 5.279 ms
     Changes:   

Summary for local
-------------
Succeeded: 28 (changed=26)
Failed:     1
```

File: dom0/sd-workstation.top [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
```
Comment (emkll) [inline dom0/sd-workstation.top]: This is a massive improvement

File: scripts/destroy-vm [hunk 4]
```diff
@@ -1,37 +1,82 @@
...
+def destroy_vm(vm):
+    """
+    Destroys a single VM instance. Requires arg to be
+    QubesVM object.
+    """
+    assert SDW_DEFAULT_TAG in vm.tags
+    if vm.is_running():
+        vm.kill()
+    print("Destroying VM '{}'... ".format(vm.name), end="")
+    subprocess.check_call("qvm-remove -f {}".format(vm.name).split())
```
Comment (emkll) [inline scripts/destroy-vm]: nit: why not explicitly write the array here? probably less overall characters and easier to read

File: scripts/destroy-vm [hunk 5]
```diff
@@ -1,37 +1,82 @@
...
+
+def destroy_all():
+    """
+    Destroys all VMs marked with the 'sd-workstation' tag, in the following order:
+    DispVMs, AppVMs, then TemplateVMs. Excludes VMs for which
+    installed_by_rpm=true.
+    """
+    # Remove DispVMs first, then AppVMs, then TemplateVMs last.
+    sdw_vms = [vm for vm in q.domains if SDW_DEFAULT_TAG in vm.tags]
+    sdw_template_vms = [vm for vm in sdw_vms
```
Comment (emkll) [inline scripts/destroy-vm]: This is very helpful, let's refactor/reuse this in https://github.com/freedomofpress/securedrop-workstation/pull/352

File: dom0/fpf-apt-test-repo.sls [hunk 6]
```diff
@@ -1,19 +1,8 @@
...
-    # Unclear why "Delete" *must* be capitalized, but that's the case!
-    - mode: delete
-    - match: jessie-backports
-    # quiet param seems to be ignored, so using "onlyif" to test existence
-    - quiet: True
-    - onlyif:
-      - test -f /etc/apt/sources.list
+#
+include:
+  - update.qubes-vm
```
Comment (emkll) [inline dom0/fpf-apt-test-repo.sls]: Where is this salt file located? Is this provided by qubes' build in salt mgmt commands? Might be worth adding a comment here for future maintainers.

File: dom0/fpf-apt-test-repo.sls [hunk 7]
```diff
@@ -22,22 +11,14 @@ install-python-apt-for-repo-config:
     - pkgs:
       - python-apt
     - require:
-      - file: remove-jessie-backports-repo
+      # Require that the Qubes update state has run first
```
Comment (emkll) [inline dom0/fpf-apt-test-repo.sls]: same comment as https://github.com/freedomofpress/securedrop-workstation/pull/351/files#diff-f739e903ed54a409c69c5835698a9336R5

**Comment by emkll** [comment]
It just occurred to me, as I am reading Qubes Fedora-29 EOL announcement [1], that even with the changes introduced here, we not remove unused RPM template in dom0: but we would need to check if any AppVMs are based on that template, otherwise the dnf remove operation would fail.

This 

[1] https://www.qubes-os.org/news/2019/11/29/fedora-29-eol/

File: dom0/sd-clean-all.sls [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+{% if sdw_customized_rpc_files|length > 0 %}
+sd-cleanup-rpc-policy-grants:
+  file.replace:
+    - names: {{ sdw_customized_rpc_files }}
+    - pattern: '### BEGIN securedrop-workstation ###.*### END securedrop-workstation ###'
+    - flags:
+      - MULTILINE
+      - DOTALL
+    - repl: ''
```
Comment (emkll) [inline dom0/sd-clean-all.sls]: While running make all/make clean several times, I've observed blank lines in the rpc policy files

I think that adding `append_newline` to `False` here could resolve

File: dom0/sd-clean-all.sls (hunk 1 above)
Comment (conorsch) [reply dom0/sd-clean-all.sls]: These policies are created automatically during upgrades via a disp-mgmt VM, and also removed automatically. If someone ctrl+c's the `make all` process, it's possible for them to hang around. Possibly overkill to worry about them here, I'm happy to excise.

File: dom0/sd-clean-all.sls (hunk 2 above)
Comment (conorsch) [reply dom0/sd-clean-all.sls]: Hmm, great catch. We can just pass `-I` to grep so it doesn't report binary files. That should handle the edge case you describe.

File: scripts/destroy-vm (hunk 4 above)
Comment (conorsch) [reply scripts/destroy-vm]: Fair, that's a rather inscrutable line as written. :smiley: Will update as you suggest!

File: dom0/fpf-apt-test-repo.sls (hunk 6 above)
Comment (conorsch) [reply dom0/fpf-apt-test-repo.sls]: +1 for more comments everywhere, will add.

File: dom0/sd-clean-all.sls (hunk 8 above)
Comment (conorsch) [reply dom0/sd-clean-all.sls]: Interesting, have not seen that, will get a repro and cleanup as you suggest.

**Comment by conorsch** [comment]
@emkll All requested changes have been addressed. Ready for re-review.

> we not remove unused RPM template in dom0: but we would need to check if any AppVMs are based on that template, otherwise the dnf remove operation would fail.

Agreed, we shouldn't be purging the old RPMs until we're sure the operation will succeed. As we make heavier use of the Admin API, we can implement "migration" tasks that handle each case intelligently, including references to DispVM settings. Both are relevant for #352, will review with that in mind.

**Pull request merged**